sorcerer/src/sorcerer.rs

```diff
@@ -0,0 +1,182 @@
+
+use futures::{FutureExt, StreamExt};
+use libp2p::PeerId;
+use maplit::hashmap;
+
+use aquamarine::AquamarineApi;
+use connection_pool::ConnectionPoolApi;
+use events_dispatcher::scheduler::api::{Event, SchedulerApi};
+use fluence_libp2p::types::Inlet;
+use kademlia::KademliaApi;
+use particle_builtins::{wrap, wrap_unit, Builtins};
+use particle_execution::ServiceFunction;
+use particle_modules::ModuleRepository;
+use particle_services::ParticleAppServices;
+use server_config::ResolvedConfig;
+use spell_storage::SpellStorage;
+
+#[derive(Clone)]
+pub struct Sorcerer {
+    pub aquamarine: AquamarineApi,
+    pub services: ParticleAppServices,
+    pub spell_storage: SpellStorage,
+    pub scheduler_api: SchedulerApi,
+    /// it is temporary, later we will use spell keypairs
+    pub node_peer_id: PeerId,
+}
+
+impl Sorcerer {
+    pub fn new<C>(
+        builtins: Arc<Builtins<C>>,
+        aquamarine: AquamarineApi,
+        config: ResolvedConfig,
+        local_peer_id: PeerId,
+        scheduler_api: SchedulerApi,
+    ) -> (Self, Vec<(String, HashMap<String, ServiceFunction>)>)
+    where
+        C: Clone + Send + Sync + 'static + AsRef<KademliaApi> + AsRef<ConnectionPoolApi>,
+    {
+        let spell_storage = Self::restore_spells(
+            config.dir_config.spell_base_dir.clone(),
+            &builtins.services,
+            &builtins.modules,
+        );
+
+        let sorcerer = Self {
+            aquamarine,
+            services: builtins.services.clone(),
+            spell_storage,
+            scheduler_api,
+            node_peer_id: local_peer_id,
+            // events_recv,
+        };
+
+        let spell_service_functions = sorcerer.get_spell_service_functions();
+
+        (sorcerer, spell_service_functions)
+    }
+
+    pub fn start(self, spell_events_stream: Inlet<Event>) -> JoinHandle<()> {
+        spawn(async {
+            spell_events_stream
+                .for_each_concurrent(None, move |event| {
+                    let sorcerer = self.clone();
+                    match event {
+                        Event::TimeTrigger { id } => {
+                            async move {
+                                sorcerer.execute_script(id).await;
+                            }
+                        }
+                    }
+                })
+                .await;
+        })
+    }
+
+    fn get_spell_service_functions(&self) -> Vec<(String, HashMap<String, ServiceFunction>)> {
+        let mut service_functions: Vec<(String, HashMap<String, ServiceFunction>)> = vec![];
+        let services_install = self.services.clone();
+        let storage_install = self.spell_storage.clone();
+        let scheduler_api_install = self.scheduler_api.clone();
+        let install_closure: ServiceFunction = Box::new(move |args, params| {
+            let storage = storage_install.clone();
+            let services = services_install.clone();
+            let scheduler_api = scheduler_api_install.clone();
```

Comment:
Not sure why you added `_install`, but it hurts my eyes

```suggestion
        let mut service_functions: Vec<(String, HashMap<String, ServiceFunction>)> = vec![];
        let services = self.services.clone();
        let storage = self.spell_storage.clone();
        let scheduler = self.scheduler_api.clone();
        let install_closure: ServiceFunction = Box::new(move |args, params| {
            let services = services.clone();
            let storage = storage.clone();
            let scheduler_api = scheduler.clone();
```